File: src/main/java/ru/nucodelabs/gem/view/PowerOf10Formatter.java
package ru.nucodelabs.gem.view;

import javafx.util.StringConverter;

import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.ArrayList;
import java.util.stream.Collectors;

/**
 * Formats number to ten with power of this number, only if its integer
 */
public class PowerOf10Formatter extends StringConverter<Number> {

    private static String toUpperIndex(String doubleString) {
        ArrayList<Character> resChars = new ArrayList<>();
        for (int i = 0; i < doubleString.length(); i++) {
            char c = doubleString.charAt(i);
            switch (c) {
                case '1' -> resChars.add('¹');
                case '2' -> resChars.add('²');
                case '3' -> resChars.add('³');
                case '4' -> resChars.add('⁴');
                case '5' -> resChars.add('⁵');
                case '6' -> resChars.add('⁶');
                case '7' -> resChars.add('⁷');
                case '8' -> resChars.add('⁸');
                case '9' -> resChars.add('⁹');
                case '0' -> resChars.add('⁰');
                case '.' -> resChars.add('\u0387');
                case '-' -> resChars.add('\u207b');
                default -> resChars.add(c);
            }
        }
        return resChars
                .stream()
                .map(String::valueOf)
                .collect(Collectors.joining());
    }

    @Override
    public String toString(Number object) {
        DecimalFormat format = new DecimalFormat();
        DecimalFormatSymbols formatSymbols = new DecimalFormatSymbols();
        formatSymbols.setDecimalSeparator('.');
        format.setDecimalFormatSymbols(formatSymbols);
        if (object.doubleValue() == 0) {
            return "1";
        } else if (object.doubleValue() - Math.ceil(object.doubleValue()) == 0) {
            return ("10" + toUpperIndex(format.format(object.doubleValue())));
        } else {
            return "";
        }
    }

    @Override
    public Number fromString(String string) {
        return null;
    }
}

File: src/main/java/ru/nucodelabs/gem/view/PowerOf10Formatter.kt
package ru.nucodelabs.gem.view

import javafx.util.StringConverter
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.util.Locale

/**
 * Formats numbers as powers of ten when the value is integer.
 */
class PowerOf10Formatter : StringConverter<Number>() {

    private fun toUpperIndex(value: String): String = buildString {
        value.forEach { c ->
            append(
                when (c) {
                    '1' -> '¹'
                    '2' -> '²'
                    '3' -> '³'
                    '4' -> '⁴'
                    '5' -> '⁵'
                    '6' -> '⁶'
                    '7' -> '⁷'
                    '8' -> '⁸'
                    '9' -> '⁹'
                    '0' -> '⁰'
                    '.' -> '\u0387'
                    '-' -> '\u207b'
                    else -> c
                }
            )
        }
    }

    override fun toString(number: Number): String {
        val format = DecimalFormat().apply {
            decimalFormatSymbols = DecimalFormatSymbols(Locale.US).apply {
                decimalSeparator = '.'
            }
        }
        return when {
            number.toDouble() == 0.0 -> "1"
            number.toDouble() - Math.ceil(number.toDouble()) == 0.0 ->
                "10" + toUpperIndex(format.format(number.toDouble()))
            else -> ""
        }
    }

    override fun fromString(string: String?): Number? {
        throw UnsupportedOperationException()
    }
}
